### ISSUE TITLE
Why does performance differ systematically?

### ISSUE THREAD

**Issue by bluenote10** [bheisler/criterion.rs#383]
This problem has confused me for days, making wrong performance conclusions. I was fixing a few bugs in an algorithm with the goal of a minimal performance impact. However on some code changes my code suddenly got a few percent slower, other times it "randomly" got faster again. Statistically, the outliers couldn't explain the difference, so I trusted the benchmarks.

After getting more and more confused by random performance jumps I spend a longer time repeating the measurements without any changes, and I finally saw the pattern: My benchmarks basically execute at two different "speeds" (no load on the machine, no memory effects involved). For instance one of the benchmarks either takes ~975 us or ~1000 us and the difference cannot be explained by statistical errors.

Distribution of a "fast" execution:
![image](https://user-images.githubusercontent.com/3620703/75388350-161d5e80-58e5-11ea-8312-af4ac7a957bf.png)

Distribution of a "slow" execution:
![image](https://user-images.githubusercontent.com/3620703/75388421-3cdb9500-58e5-11ea-9099-864c31fee18b.png)

Yes, there are a few outliers, but essentially the entire distribution is shifted in the second case. The reported "regression" is ~2.5 %.

My conclusion here is probably wrong.

~~Looking closer at the plots, I think the reason is the internal choice of the number of iterations. In the fast case, the bencher chose to go up to 200 iterations, compared to 100 in the slow run. It probably makes sense that doubling the iterations gives better performance because of some kind of internal warm-up.~~

~~What's the best way to deal with the problem? Maybe the comparison should never be done across benchmarks executed with a different number of iterations, i.e., if there is an existing baseline, the benchmark should use the same iteration number.~~

**Comment by bluenote10** [comment]
After more digging it looks my conclusion is wrong: I'm setting a measurement time of 1 ms now, which basically enforces that the test runs with `d = 1` every time, and yet I'm seeing systematically different performances.

As another test, I'm now doing the same measurement repeatedly directly one after the other. My benchmark code looks like:

```rust
{
    let (_, p1, p2) = load_test_case("fixtures/generic_test_cases/issue96.geojson");
    c.bench_function("issue96 / intersection 1", |b| b.iter(
        || p1.intersection(&p2),
    ));
}
{
    let (_, p1, p2) = load_test_case("fixtures/generic_test_cases/issue96.geojson");
    c.bench_function("issue96 / intersection 2", |b| b.iter(
        || p1.intersection(&p2),
    ));
}
{
    let (_, p1, p2) = load_test_case("fixtures/generic_test_cases/issue96.geojson");
    c.bench_function("issue96 / intersection 3", |b| b.iter(
        || p1.intersection(&p2),
    ));
}
```

For the first measurement I get:

![image](https://user-images.githubusercontent.com/3620703/75420509-71c50780-5938-11ea-9968-30e3ebf1bae4.png)

The second measurement comes in at:

![image](https://user-images.githubusercontent.com/3620703/75420565-8f926c80-5938-11ea-8476-559146c06401.png)

Input and algorithm are fully deterministic. Isn't it weird that in one moment in time Criterion.rs can fairly consistently measure the performance to be X and a moment later it consistently measures something which deviates significantly from X?

Any ideas what could be causing this?

**Comment by Eh2406** [comment]
Benchmarking is a black art, and I may be way off base. But:
-- The processor is changing speed. Thermal throttling or Turbo-boosting or something.
-- Cache effects. https://github.com/ccurtsinger/stabilizer suggests that where the code lines up in the executable can matter for performance.
-- Antivirus/nozy-nabers. Some other program is using your resources some of the time.
But yes this is very odd.

**Comment by bheisler** [comment]
Yeah, what Eh2406 mentioned is pretty much what I would have said.

Power-management modes in the CPU is a fairly common reason for things like this, or interference from other processes. Processor affinity might be another. The reason why Criterion.rs does all these statistics is because benchmarks tend to be affected by any number of arbitrary and unpredictable things going on in the background. You say that your code is deterministic, but even that may not be true - for example, `std::collections::HashMap` internally seeds the hasher with a bit of noise from the system RNG to prevent predictable collisions, so you would see more collisions with one run and less with the next.

**Comment by bluenote10** [comment]
Thanks for the help!

TLDR; It was the input data! Changing to a batched iter seems to solve the problem.

-----

If you're interested, here's my long-winded journey of getting to the conclusion:

- _... where the code lines up in the executable can matter for performance_ => Interesting point, I wasn't aware of that. So it is better to focus on the case of executing one and the same binary multiple times. (I was mainly putting the same benchmark twice in the same binary to rule out per-process performance effects.)
- Antivirus => Linux
- Noisy neighbors => If the environment load is homogeneous from run to run, they should be rather reflected in statistical errors. My criterion benchmark runs for ~10 seconds. When doing two consecutive runs (with no pause in between runs) it would be a big coincidence if the system load is higher exactly during the first 10 seconds and lower during the second 10 seconds (to explain that all iterations are systematically slower in the first run vs the second run). Also I tested chain-running my benchmark both under a heavy load scenario (browser with many tabs, multiple IDEs, background video) and low load (all closed except for a terminal with a single shell), and the systematic changes were the same.
- Power-management modes in the CPU / processor changing speed => I always thought that was more of an issue on laptops, but most likely accounts for the majority of the general performance variance I'm seeing.

This basically leaves the input data as the only possible explanation. The data is basically 2 x 3 vectors of floats (to represent polygons, low vertices counts), and the algorithm only briefly touches it during an initial scan, operating on internal representations all the time. Therefore I was sure it cannot have a big systematic impact, but apparently it can.

Out of curiosity I implemented a naive benchmark that simply computes the run time of 30 batches each performing 500 iterations, printing the 30 run times of the batches to stdout. Then I have an external bash loop that calls the benchmark binary 20 times. When reading the input data individually for each batch I get (y axis is runtime in μs corresponding to the 30 batches, x axis are the 20 repeated runs):

![image](https://user-images.githubusercontent.com/3620703/75623730-62022900-5bad-11ea-8fa5-262d87f91f8c.png)

No noticeable systematic differences between runs; uncertainty is well covered by statistical errors. On the other hand changing the benchmark to read the input only once for all batches gives this:

![image](https://user-images.githubusercontent.com/3620703/75623747-852cd880-5bad-11ea-895e-fc12f5af6613.png)

The high line densities point to suspiciously low statistical errors. No wonder criterion.rs sees statistically significant changes like that.
